pandas/core/reshape.py

```diff
@@ -202,10 +202,13 @@ def get_new_columns(self):
                 return self.removed_level
 
             lev = self.removed_level
-            vals = np.insert(lev.astype('object'), 0,
+            vals = np.insert(lev.values.astype('object'), 0,
                              _get_na_value(lev.dtype.type))
 
-            return lev._shallow_copy(vals)
+            try:
+                return lev._shallow_copy(vals.astype(lev.dtype, copy=False))
+            except ValueError:  # when vals cannot be type casted to lev.dtype
+                return Index(vals, **lev._get_attributes_dict())
 
```

Comment:
I get *why* you are doing the try: except, this ONLY needs to be around the astype though. Make this a separate method